src/main/scala/is/hail/distributedmatrix/Keys.scala

```diff
@@ -0,0 +1,83 @@
+package is.hail.distributedmatrix
+
+import is.hail.annotations.Annotation
+import is.hail.expr.{JSONAnnotationImpex, Parser}
+import is.hail.expr.types.Type
+import is.hail.utils._
+import org.apache.spark.SparkContext
+import org.json4s.jackson.Serialization
+import org.json4s.{JArray, JValue}
+import org.json4s.jackson.JsonMethods.parse
+
+case class KeysImpex(typeStr: String, valuesJ: JValue)
+
+object Keys {
+  def unify(left: Option[Keys], right: Option[Keys], msg: String): Option[Keys] = {
+    (left, right) match {
+      case (Some(l), Some(r)) =>
+        l.assertSame(r, msg)
+        left
+      case (Some(_), _) => left
+      case _ => right
+    }
+  }
+  
+  def read(sc: SparkContext, uri: String): Keys = {
+    val KeysImpex(typeStr, JArray(arr)) =  sc.hadoopConfiguration.readTextFile(uri)(in =>
+      try {
+        val json = parse(in)
+        json.extract[KeysImpex]
+      } catch {
+        case e: Exception => fatal(
+          s"""corrupt or outdated matrix key data.
+             |  Recreate with current version of Hail.
+             |  Detailed exception:
+             |  ${ e.getMessage }""".stripMargin)
+      })
+    
+    val typ = Parser.parseType(typeStr)
+    val values = arr.map(JSONAnnotationImpex.importAnnotation(_, typ)).toArray
+    
+    new Keys(typ, values)
+  }
+}
+
+class Keys(val typ: Type, val values: Array[Annotation]) {
```

Comment:
This is probably the best you can do now, but keep in mind: annotations are going away eventually and this will need to be reworked.